### PR TITLE
traits: add idle trait

### PR DESCRIPTION
**Pull request by xoviat**

### CODE REVIEW

**Comment by xoviat** [comment]
Note, for stm32, `  return (READ_BIT(((DMA_Channel_TypeDef *)((uint32_t)((uint32_t)DMAx + CHANNEL_OFFSET_TAB[Channel - 1U])))->CNDTR,
                   DMA_CNDTR_NDT));`

File: embassy-stm32f4/src/serial.rs [hunk 1]
```diff
@@ -181,6 +232,8 @@ where
                 () = fut_idle.fuse() => {},
             }
 
+            let bytes = RSTREAM::get_number_of_transfers() as usize;
```
Comment (thalesfragoso) [inline embassy-stm32f4/src/serial.rs:235]: This probably gives the remaining number of transfers, not the number of transfers that were completed.

Comment (xoviat) [reply embassy-stm32f4/src/serial.rs:235]: I think it's correct, for this reason:  the registers read by this function appear to be the same ones as that read by examples written in C.

Comment (thalesfragoso) [reply embassy-stm32f4/src/serial.rs:235]: From the manual:

> Number of data items to be transferred (0 up to 65535). This register can be written only
when the stream is disabled. When the stream is enabled, this register is read-only,
**indicating the remaining data items to be transmitted**. This register decrements after each
DMA transfer.
Once the transfer has completed, this register can either stay at **zero** (when the stream is in
normal mode) or be reloaded automatically with the previously programmed value in the
following cases:
>
>– when the stream is configured in Circular mode.
>– when the stream is enabled again by setting EN bit to '1'
>
>If the value of this register is zero, no transaction can be served even if the stream is enabled.

**Comment by xoviat** [comment]
@thalesfragoso Do you think this would work?

**Comment by thalesfragoso** [comment]
>  Do you think this would work?

Probably, yes.

**Comment by xoviat** [comment]
@theunkn0wn1 Does this work for you?

File: embassy-traits/src/uart.rs [hunk 2]
```diff
@@ -10,6 +10,15 @@ pub enum Error {
 pub trait Uart {
     type ReceiveFuture<'a>: Future<Output = Result<(), Error>>;
     type SendFuture<'a>: Future<Output = Result<(), Error>>;
+    // Receive into the buffer until the buffer is full
```
Comment (theunkn0wn1) [inline embassy-traits/src/uart.rs:13]: You need to use `///` to create a docstring.
```suggestion
    /// Receive into the buffer until the buffer is full
```

File: embassy-traits/src/uart.rs [hunk 3]
```diff
@@ -10,6 +10,15 @@ pub enum Error {
...
     fn receive<'a>(&'a mut self, buf: &'a mut [u8]) -> Self::ReceiveFuture<'a>;
+    // Send the specified buffer, and return when the transmission has completed
```
Comment (theunkn0wn1) [inline embassy-traits/src/uart.rs:15]: you need to use `///` to create a docstring.
```suggestion
    /// Send the specified buffer, and return when the transmission has completed
```

File: embassy-traits/src/uart.rs [hunk 4]
```diff
@@ -10,6 +10,15 @@ pub enum Error {
...
     fn send<'a>(&'a mut self, buf: &'a [u8]) -> Self::SendFuture<'a>;
 }
+
+pub trait IdleUart {
+    type ReceiveFuture<'a>: Future<Output = Result<usize, Error>>;
+    // Receive into the buffer until the buffer is full or the line is idle after some bytes are received
+    // Return the number of bytes received
```
Comment (theunkn0wn1) [inline embassy-traits/src/uart.rs:22]: You need to use `///` to create a docstring.
```suggestion
    /// Receive into the buffer until the buffer is full or the line is idle after some bytes are received
    /// Return the number of bytes received
```

**Review by theunkn0wn1 — APPROVED** [review]
Beyond the minor docstring nits, this LGTM.  
It appears to behave correctly on my f446 as well.

File: embassy-stm32f4/src/serial.rs [hunk 5]
```diff
@@ -150,6 +150,57 @@ where
...
+    CHANNEL: Channel + 'static,
+{
+    type ReceiveFuture<'a> = impl Future<Output = Result<usize, Error>> + 'a;
+
+    /// Receives serial data.
+    ///
+    /// The future is pending until the buffer is completely filled.
+    /// A common pattern is to use [`stop()`](ReceiveFuture::stop) to cancel
+    /// unfinished transfers after a timeout to prevent lockup when no more data
+    /// is incoming.
```
Comment (theunkn0wn1) [inline embassy-stm32f4/src/serial.rs:197]: On review, this docstring is incorrect.
This future completes either when the buffer is completely filled OR the line falls idle after reading some data.
```suggestion
    /// Receives serial data.
    ///
    /// The future is pending until either the buffer is completely full, or the RX line falls idle after receiving some data.
    ///
    /// Returns the number of bytes read.
```

**Review by Dirbaio — APPROVED** [review]
:+1:  on the trait! 

I won't comment on the stm32 impl since I'm not an stm32 expert :)

**Pull request merged**